mitmproxy/net/tls.py

```diff
@@ -1,266 +1,147 @@
-
-    if options.ssl_insecure:
-        verify = SSL.VERIFY_NONE
-    else:
-        verify = SSL.VERIFY_PEER
 
-    method, tls_options = VERSION_CHOICES[options.ssl_version_server]
+# TODO: remove once https://github.com/pyca/pyopenssl/pull/985 has landed.
+SSL.Context._methods.setdefault(Method.TLS_SERVER_METHOD.value, "TLS_server_method")
+SSL.Context._methods.setdefault(Method.TLS_CLIENT_METHOD.value, "TLS_client_method")
 
-    return {
-        "verify": verify,
-        "method": method,
-        "options": tls_options,
-        "ca_path": options.ssl_verify_upstream_trusted_confdir,
-        "ca_pemfile": options.ssl_verify_upstream_trusted_ca,
-        "client_certs": options.client_certs,
-        "cipher_list": options.ciphers_server,
-    }
+
+class Version(Enum):
+    UNBOUNDED = 0
+    # TODO: just SSL attributes once https://github.com/pyca/pyopenssl/pull/985 has landed.
+    SSL3 = getattr(SSL, "SSL3_VERSION", 768)
+    TLS1 = getattr(SSL, "TLS1_VERSION", 768)
+    TLS1_1 = getattr(SSL, "TLS1_1_VERSION", 770)
+    TLS1_2 = getattr(SSL, "TLS1_2_VERSION", 771)
+    TLS1_3 = getattr(SSL, "TLS1_3_VERSION", 772)
+
+
+class Verify(Enum):
+    VERIFY_NONE = SSL.VERIFY_NONE
+    VERIFY_PEER = SSL.VERIFY_PEER
+
+
+DEFAULT_MIN_VERSION = Version.TLS1_2
+DEFAULT_MAX_VERSION = Version.UNBOUNDED
+DEFAULT_OPTIONS = (
+        SSL.OP_CIPHER_SERVER_PREFERENCE
+        | SSL.OP_NO_COMPRESSION
+)
 
 
 class MasterSecretLogger:
-    def __init__(self, filename):
-        self.filename = os.path.expanduser(filename)
+    def __init__(self, filename: Path):
+        self.filename = filename.expanduser()
         self.f = None
         self.lock = threading.Lock()
 
     # required for functools.wraps, which pyOpenSSL uses.
     __name__ = "MasterSecretLogger"
 
-    def __call__(self, connection, keymaterial):
+    def __call__(self, connection: SSL.Connection, keymaterial: bytes) -> None:
         with self.lock:
-            if not self.f:
-                d = os.path.dirname(self.filename)
-                if not os.path.isdir(d):
-                    os.makedirs(d)
-                self.f = open(self.filename, "ab")
+            if self.f is None:
+                self.filename.parent.mkdir(parents=True, exist_ok=True)
+                self.f = self.filename.open("ab")
                 self.f.write(b"\n")
-            self.f.write(keymaterial)
-            self.f.write(b"\n")
+            self.f.write(keymaterial + b"\n")
             self.f.flush()
 
     def close(self):
         with self.lock:
-            if self.f:
+            if self.f is not None:
                 self.f.close()
 
-    @staticmethod
-    def create_logfun(filename):
-        if filename:
-            return MasterSecretLogger(filename)
-        return None
+
+def make_master_secret_logger(filename: Optional[str]) -> Optional[MasterSecretLogger]:
+    if filename:
+        return MasterSecretLogger(Path(filename))
+    return None
```

Comment:
Just for fun: this might be a good place for a Null-Object:
https://en.wikipedia.org/wiki/Null_object_pattern

Reply:
You mean having a noop function? I decided against that because we can avoid the entire C -> Python -> C roundtrip, which probably isn't too cheap.